README.md
The web page of this program was originally hosted by geocities.jp,
which will be closed by March 2019.  
Here is a copy of that to keep it accessible in the future.  
No longer maintained.

    This is formatting disk utility for USB floppy devices.
    
    Usage: ufiformat [OPTION]... [DEVICE]
    Format a floppy disk in a USB floppy disk DEVICE.
    
      -f, --format [SIZE]  specify format capacity SIZE in KB
                           without -f option, the format of the current media will be used
      -F, --force          do not perform any safety checks
      -i, --inquire        show device information, instead of performing format
                           without DEVICE argument, list USB floppy disk devices
      -v, --verbose        show detailed output
      -q, --quiet          suppress minor output
      -h, --help           show this message
    
    NOTE: ufiformat supports only the following format capacities.
             1440/1232/1200 (for 2hd disk)
             720/640        (for 2dd disk)
          The device should support the capacities also,
          otherwise ufiformat shows an error message.
    
          The above format capacities are predefined in the program, but
          each usb floppy device also has a limited set of formats (defined internally)
          that it can format media to.  The allowed format capacities are obtained by
          querying the device, but this only returns the total format capacity
          and not CHS (cylinder, heads and sectors), hence a mapping is
          required in the program.
    
    This program is based around the following document:
      "Universal Serial Bus Mass Storage Class - UFI Command Specification"
      Revision 1.0 December 14 1998
      http://www.usb.org/developers/devclass_docs/usbmass-ufi10.pdf
    
    Currently it is known to work under the following environment.
      * Plamo Linux 4.0.2(kernel 2.6) + YE-DATA USB-FDU(OEM)
      * Plamo Linux 4.0.2(kernel 2.4) + YE-DATA USB-FDU(OEM)
      * SuSE-10 + TEAC FD-05PUB(OEM)
      * Linux Zaurus SL-C700(kernel 2.4) + REX-CFU1 + YE-DATA USB-FDU
        (some kernel modules and recent libext2 are needed)


ChangeLog
2011-Jan-29  Kazuhiro Hayashi  <tedi@tedi.sakura.ne.jp>

	Version 0.9.9
	* support for old sysfs, patch contributed by Kabe
	* add spec file, also contributed by Kabe

2010-Nov-29  Kazuhiro Hayashi  <tedi@tedi.sakura.ne.jp>

	Version 0.9.8
	* verification support for some unusual devices
        * fix typo

2010-Nov-21  Kazuhiro Hayashi  <tedi@tedi.sakura.ne.jp>

	Version 0.9.7
	* add verification feature, thanks to Robert M. Stockmann

2009-Nov-02  Kazuhiro Hayashi  <tedi@tedi.sakura.ne.jp>

	Version 0.9.6
	* bug fix of the method of usb floppy detection with sysfs

2009-Oct-03  Kazuhiro Hayashi  <tedi@tedi.sakura.ne.jp>

	Version 0.9.5
	* use sysfs if it exists

2007-Nov-10  Kazuhiro Hayashi  <tedi@tedi.sakura.ne.jp>

	Version 0.9.4
	* add path info in a error message when /dev/sg* is unavailable
	  suggested by <kabe@sra-tohoku.co.jp>
	* add manpage
	  contributed by <kabe@sra-tohoku.co.jp>
	* store ufiformat binary to (prefix)/bin instead of sbin

2006-Nov-25  Kazuhiro Hayashi  <tedi@tedi.sakura.ne.jp>

	Version 0.9.3
	* bug fix for device detection
	  patch contributed by Martin <martint@a1.net>

2006-Mar-08  Kazuhiro Hayashi  <tedi@tedi.sakura.ne.jp>

	Version 0.9.2(beta release)
	* show error with device path if possible
	* small bug fix
	* add some documents

2006-Feb-19  Kazuhiro Hayashi  <tedi@tedi.sakura.ne.jp>

	Version 0.9.1
	* show error messages properly

2006-Feb-19  Kazuhiro Hayashi  <tedi@tedi.sakura.ne.jp>

	Version 0.9.0
	* total refactoring for the release
	* clean up of command line options
	* add some safety check
	* modify to work under kernel-2.4/2.6 (using procfs)
	
2005-Dec-15  John Floyd  <jfloyd@bigpond.net.au>

	Version 0.3
	* Small cleanups of output.  Free all sysfs lists.
	* Added usage, -h/--help to options.
	* Changed program name to ufiformat.

2005-Dec-14  John Floyd  <jfloyd@bigpond.net.au>

	Version 0.2.2
	* Fix readcapacities to allow an non-ready drive to move into an 
	  active state. Loop around the READCAPACITIES sg command until 
	  ready.

2005-Dec-11  John Floyd  <jfloyd@bigpond.net.au>

	Version 0.2.1
	* Test for null return on sysfs attrib query (in 
          ufi_sysfs:get_floppy_devices).
	* Return no devices found (in ufi_sysfs:floppy_list_print)

2005-Dec-10  John Floyd  <jfloyd@bigpond.net.au>

	Version 0.2
	* Added SySFS Scan to determine ufi capable devices (ufi_sysfs.c)
	* Added SySFS Scan to map block devices to generic scsii device.
	* Added Read_Format_Capacities (ufi command) to determine allowed formats.
	  (ufi_read_capacities.c)
	* using autoconf & automake

2005-Jan-14  Kazuhiro Hayashi  <tedi@tedi.sakura.ne.jp>

	* initial version


ufi_detect.c
/*
    ufiformat Version 0.9.9 2011/01/29

    Copyright (C) 2005-2011 Kazuhiro Hayashi <tedi@tedi.sakura.ne.jp>
    Copyright (C) 2005 John Floyd <jfloyd@bigpond.net.au>

    The method of formatting a floppy on USB-FDD used in this program
    is introduced by Bruce M Simpson.

    This program is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public License
    as published by the Free Software Foundation; either version 2
    of the License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
*/

#ifdef HAVE_CONFIG_H
#  include "config.h"
#endif
#include <sys/types.h>
#include <sys/param.h>
#include <sys/stat.h>
#include <sys/sysmacros.h>
#include <linux/major.h>
#include <fcntl.h>
#include <scsi/scsi.h>
#include <sys/ioctl.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <errno.h>

#include "ufi_detect.h"

#ifndef SCSI_DISK_MAJOR
#define SCSI_DISK_MAJOR(M) ((M) == SCSI_DISK0_MAJOR || \
        ((M) >= SCSI_DISK1_MAJOR && (M) <= SCSI_DISK7_MAJOR) || \
        ((M) >= SCSI_DISK8_MAJOR && (M) <= SCSI_DISK15_MAJOR))
#endif

#define HOST_ID(idlun) ((idlun >> 24) & 0xff)

static char *pathcpy(char *dst, const char *src, size_t len)
{
    if (strlen(src) >= len) {
	errno = ENAMETOOLONG;
	return NULL;
    }
    return strcpy(dst, src);
}

static int is_sd_device(const char *dev)
{
    struct stat st;

    if (stat(dev, &st) < 0) {
	return -1;
    }
    return S_ISBLK(st.st_mode) && SCSI_DISK_MAJOR(major(st.st_rdev)) && (minor(st.st_rdev) & 0xf) == 0 ? 1 : 0;
}

static int is_sg_device(const char *dev)
{
    struct stat st;

    if (stat(dev, &st) < 0) {
	return -1;
    }
    return S_ISCHR(st.st_mode) && major(st.st_rdev) == SCSI_GENERIC_MAJOR ? 1 : 0;
}

static int get_id_lun(const char *dev)
{
    int fd;
    int idlun[2];

    if ((fd = open(dev, O_RDONLY | O_NONBLOCK)) < 0) {
	return -1;
    }
    if (ioctl(fd, SCSI_IOCTL_GET_IDLUN, idlun) < 0) {
	close(fd);
	return -1;
    }
    close(fd);
    return idlun[0];
}

static int get_sd_path(int idlun, char *path, size_t len, int host_id_only)
{
    char tmp[PATH_MAX];
    int i;
    int eacces;
    int idlun2;

    eacces = 0;
    /* BUG: only single usb fdd is supported for one host_id */
    if (host_id_only) {
	idlun = (idlun & 0xff) << 24;
    }
    for (i = 0; i < 256; i++) {
	if (i < 26) {
	    sprintf(tmp, "/dev/sd%c", 'a' + i);
	} else {
	    sprintf(tmp, "/dev/sd%c%c", 'a' + (i / 26), 'a' + (i % 26));
	}
	if ((idlun2 = get_id_lun(tmp)) == -1) {
	    if (errno == EACCES) {
		eacces = 1;
	    }
	} else if (idlun == idlun2) {
	    if (pathcpy(path, tmp, len) == NULL) {
		return -1;
	    }
 	    return 0;
	}
    }
    errno = eacces ? EACCES : ENOENT;
    return -1;
}

static int get_sg_path(int idlun, char *path, size_t len, int host_id_only)
{
    char tmp[PATH_MAX];
    int i;
    int eacces;
    int idlun2;

    eacces = 0;
    /* BUG: only single usb fdd is supported for one host_id */
    if (host_id_only) {
	idlun = (idlun & 0xff) << 24;
    }
    for (i = 0; i < 256; i++) {
	sprintf(tmp, "/dev/sg%d", i);
	if ((idlun2 = get_id_lun(tmp)) == -1) {
	    if (errno == EACCES) {
		eacces = 1;
	    }
	} else if (idlun == idlun2) {
	    if (pathcpy(path, tmp, len) == NULL) {
		return -1;
	    }
 	    return 0;
	}
    }
    errno = eacces ? EACCES : ENOENT;
    return -1;
}

int convert_to_sg_path(const char *sd, char *sg, size_t len)
{
    int idlun;
    int flag;

    if ((flag = is_sg_device(sd)) < 0) {
	return -1;
    }
    if (flag) {
	if (pathcpy(sg, sd, len) == NULL) {
	    return -1;
	}
	return 0;
    }
    if ((flag = is_sd_device(sd)) < 0) {
	return -1;
    }
    if (!flag) {
	errno = ENODEV;
	return -1;
    }
    if ((idlun = get_id_lun(sd)) == -1) {
	return -1;
    }
    if (get_sg_path(idlun, sg, len, 0) < 0) {
	errno = -errno;
	return -1;
    }
    return 0;
}

int convert_to_sd_path(const char *sg, char *sd, size_t len)
{
    int idlun;
    int flag;

    if ((flag = is_sd_device(sg)) < 0) {
	return -1;
    }
    if (flag) {
	if (pathcpy(sd, sg, len) == NULL) {
	    return -1;
	}
	return 0;
    }
    if ((flag = is_sg_device(sg)) < 0) {
	return -1;
    }
    if (!flag) {
	errno = ENODEV;
	return -1;
    }
    if ((idlun = get_id_lun(sg)) == -1) {
	return -1;
    }
    if (get_sd_path(idlun, sd, len, 0) < 0) {
	errno = -errno;
	return -1;
    }
    return 0;
}

static int is_usb_fdd_for_host(int host_id)
{
    int i;
    char path[PATH_MAX];
    char buf[1024];
    FILE *fp;
    struct stat st;

    sprintf(path, "/sys/class/scsi_host/host%d/device/../driver", host_id);
    if (stat(path, &st) == 0) {
	int n;

	if ((n = readlink(path, buf, sizeof(buf))) < 12 || strncmp(buf + n - 12, "/usb-storage", 12) != 0) {
	    return 0;
	}

	sprintf(path, "/sys/class/scsi_host/host%d/device/../bInterfaceClass", host_id);
	if ((fp = fopen(path, "r")) == NULL) {
	    return 0;
	}
	if (fgets(buf, sizeof(buf), fp) == NULL || strcmp(buf, "08\n") != 0) {
	    fclose(fp);
	    return 0;
	}
	fclose(fp);

	sprintf(path, "/sys/class/scsi_host/host%d/device/../bInterfaceSubClass", host_id);
	if ((fp = fopen(path, "r")) == NULL) {
	    return 0;
	}
	if (fgets(buf, sizeof(buf), fp) == NULL || strcmp(buf, "04\n") != 0) {
	    fclose(fp);
	    return 0;
	}
	fclose(fp);

	return 1;
    }

    sprintf(path, "/proc/scsi/usb-storage/%d", host_id);
    if ((fp = fopen(path, "r")) == NULL) {
	for (i = 0; i < 256; i++) {
	    sprintf(path, "/proc/scsi/usb-storage-%d/%d", i, host_id);
	    if ((fp = fopen(path, "r")) != NULL) {
		break;
	    }
	}
	if (fp == NULL) {
	    return 0;
	}
    }
    while (fgets(buf, sizeof(buf), fp) != NULL) {
	if (strcmp(buf, "     Protocol: Uniform Floppy Interface (UFI)\n") == 0) {
	    fclose(fp);
	    return 1;
	}
    }
    fclose(fp);
    return 0;
}

int is_usb_fdd(const char *dev)
{
    int idlun;
    int is_sd;
    int is_sg;

    if ((is_sd = is_sd_device(dev)) < 0 || (is_sg = is_sg_device(dev)) < 0) {
	return -1;
    }
    if (!is_sd && !is_sg) {
	return 0;
    }
    if ((idlun = get_id_lun(dev)) == -1) {
	return -1;
    }
    return is_usb_fdd_for_host(HOST_ID(idlun));
}

int get_usb_fdds(struct ufi_path_info **info)
{
    int i;
    struct ufi_path_info *top = NULL;
    struct ufi_path_info *last = NULL;
    struct ufi_path_info *current;
    char sg_path[PATH_MAX];
    char sd_path[PATH_MAX];

    for (i = 0; i < 256; i++) {
	if (is_usb_fdd_for_host(i)) {
	    if (get_sg_path(i, sg_path, sizeof(sg_path), 1) < 0) {
		if (errno != ENOENT) {
		    free_ufi_path_info(top);
		    return -1;
		}
		sg_path[0] = '\0';
	    }
	    if (get_sd_path(i, sd_path, sizeof(sd_path), 1) < 0) {
		if (errno != ENOENT) {
		    free_ufi_path_info(top);
		    return -1;
		}
		sd_path[0] = '\0';
	    }
	    current = malloc(sizeof(struct ufi_path_info));
	    current->host_id = i;
	    current->sg_path = sg_path[0] != '\0' ? strdup(sg_path) : NULL;
	    current->sd_path = sd_path[0] != '\0' ? strdup(sd_path) : NULL;
	    current->next = NULL;
	    if (last != NULL) {
		last->next = current;
	    }
	    if (top == NULL) {
		top = current;
	    }
	    last = current;
	}
    }
    *info = top;
    return 0;
}

void free_ufi_path_info(struct ufi_path_info *info)
{
    while (info != NULL) {
	if (info->sg_path != NULL) {
	    free(info->sg_path);
	}
	if (info->sd_path != NULL) {
	    free(info->sd_path);
	}
	info = info->next;
    }
}


ufi_detect.h
/*
    ufiformat Version 0.9.9 2011/01/29

    Copyright (C) 2005-2011 Kazuhiro Hayashi <tedi@tedi.sakura.ne.jp>
    Copyright (C) 2005 John Floyd <jfloyd@bigpond.net.au>

    The method of formatting a floppy on USB-FDD used in this program
    is introduced by Bruce M Simpson.

    This program is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public License
    as published by the Free Software Foundation; either version 2
    of the License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
*/

#ifndef UFI_DETECT_H
#define UFI_DETECT_H

struct ufi_path_info {
    int host_id;
    char *sg_path;
    char *sd_path;
    struct ufi_path_info *next;
};

int convert_to_sg_path(const char *sd, char *sg, size_t len);
int convert_to_sd_path(const char *sg, char *sd, size_t len);
int is_usb_fdd(const char *dev);
int get_usb_fdds(struct ufi_path_info **info);
void free_ufi_path_info(struct ufi_path_info *info);

#endif



ufi_command.h
/*
    ufiformat Version 0.9.9 2011/01/29

    Copyright (C) 2005-2011 Kazuhiro Hayashi <tedi@tedi.sakura.ne.jp>
    Copyright (C) 2005 John Floyd <jfloyd@bigpond.net.au>

    The method of formatting a floppy on USB-FDD used in this program
    is introduced by Bruce M Simpson.

    This program is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public License
    as published by the Free Software Foundation; either version 2
    of the License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
*/

#ifndef UFI_COMMAND_H
#define UFI_COMMAND_H

struct ufi_capacities {
    int blocks;
    int block_size;
    struct ufi_capacities *next;
};

struct ufi_product {
    char *vendor;
    char *product;
};

void ufi_set_verbose(int verbose);
int ufi_test_unit_ready(int fd);
struct ufi_capacities *ufi_read_format_capacities(int fd, int *type);
struct ufi_product *ufi_inquiry(int fd);
int ufi_mode_sense(int fd, int *write_protected);
int ufi_format_unit(int fd, int blocks, int block_size, int track, int head);
int ufi_reset(int fd);

#define UFI_GOOD 0
#define UFI_ERROR -1
#define UFI_UNFORMATTED_MEDIA 1
#define UFI_FORMATTED_MEDIA 2
#define UFI_NO_MEDIA 3

#define UFI_PROTECTED 1
#define UFI_NOT_PROTECTED 0
#endif


ufi_format.c
/*
    ufiformat Version 0.9.9 2011/01/29

    Copyright (C) 2005-2011 Kazuhiro Hayashi <tedi@tedi.sakura.ne.jp>
    Copyright (C) 2005 John Floyd <jfloyd@bigpond.net.au>

    The method of formatting a floppy on USB-FDD used in this program
    is introduced by Bruce M Simpson.

    This program is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public License
    as published by the Free Software Foundation; either version 2
    of the License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
*/

#ifdef HAVE_CONFIG_H
#  include "config.h"
#endif
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <sys/types.h>
#include <sys/param.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <errno.h>
#include <ext2fs/ext2fs.h>
#include "ufi_detect.h"
#include "ufi_command.h"

#ifndef PACKAGE
#define PACKAGE "ufiformat"
#endif

static void list_devices(int verbose)
{
    struct ufi_path_info *path_info_p;

    if (verbose > 1) {
	printf("list usb fdd devices\n");
    }
    if (verbose) {
	printf("%-12s %-12s\n", "disk", "generic");
    }
    if (get_usb_fdds(&path_info_p) < 0) {
	goto error;
    }
    while (path_info_p != NULL) {
	if (path_info_p->sd_path != NULL || path_info_p->sg_path != NULL) {
	    printf("%-12s %-12s\n",
		   path_info_p->sd_path != NULL ? path_info_p->sd_path : "",
		   path_info_p->sg_path != NULL ? path_info_p->sg_path : "");
	} else {
	    if (verbose > 1) {
		printf("ignoreing unattached fdd: host_id = %d\n", path_info_p->host_id);
	    }
	}
	path_info_p = path_info_p->next;
    }
    return;

 error:
    perror("");
    exit(1);
}

static void show_capacity(const char *status, int blocks, int block_size)
{
    printf("%-12s %4d %4d %4d\n", status, blocks, block_size, blocks * block_size / 1024);
}

#define IS_2HD(size) ((size) > 1024 ? 1 : 0)

static void inquire(const char *path, int verbose, int force)
{
    char *sg = NULL;
    int fd = -1;
    struct ufi_product *product;
    struct ufi_capacities *capacities;
    int type;
    int wp;
    int is_fdd;

    if (verbose > 1) {
	printf("inquire on device=%s\n", path);
    }
    if (!force) {
	if ((is_fdd = is_usb_fdd(path)) < 0) {
	    goto error;
	}
	if (!is_fdd) {
	    fprintf(stderr, "%s: device is not usb fdd\n", path);
	    goto error2;
	}
	sg = malloc(PATH_MAX);
	if (convert_to_sg_path(path, sg, PATH_MAX)) {
	    if (verbose > 1 && errno == -ENOENT) {
		fprintf(stderr, "\"modprobe sg\" might be needed to prepare SCSI pass-thru device.\n");
	    }
	    sg = errno < 0 ? "/dev/sg* family" : NULL;
	    errno = errno < 0 ? -errno : errno;
	    goto error;
	}
    } else {
	sg = strdup(path);
    }
    if ((fd = open(sg, O_RDWR | O_NONBLOCK)) < 0) {
	goto error;
    }
    if (ufi_test_unit_ready(fd) == UFI_ERROR && !force) {
	goto error;
    }
    if ((capacities = ufi_read_format_capacities(fd, &type)) == NULL) {
	goto error;
    }
    if (type == UFI_UNFORMATTED_MEDIA || type == UFI_FORMATTED_MEDIA) {
	if (ufi_mode_sense(fd, &wp) != UFI_GOOD) {
	    goto error;
	}
    }
    if ((product = ufi_inquiry(fd)) == NULL) {
	goto error;
    }

    if (verbose) {
	printf("vendor:  %s\n", product->vendor);
	printf("product: %s\n", product->product);
	if (type == UFI_UNFORMATTED_MEDIA || type == UFI_FORMATTED_MEDIA) {
	    printf("write protect: %s\n", wp == UFI_PROTECTED ? "on" : "off");
	    printf("media type: %s\n", IS_2HD(capacities->blocks * capacities->block_size / 1024) ? "2HD" : "2DD");
	}
	printf("status      block size   kb\n");
    }
    switch (type) {
    case UFI_UNFORMATTED_MEDIA:
	show_capacity("unformatted", capacities->blocks, capacities->block_size);
	break;
    case UFI_FORMATTED_MEDIA:
	show_capacity("formatted", capacities->blocks, capacities->block_size);
	while (capacities->next != NULL) {
	    capacities = capacities->next;
	    show_capacity("formattable", capacities->blocks, capacities->block_size);
	}
	break;
    case UFI_NO_MEDIA:
	show_capacity("no", capacities->blocks, capacities->block_size);
	break;
    default:
	errno = EPROTO;
	goto error;
    }
    close(fd);
    return;

 error:
    perror(sg == NULL ? path : sg);
 error2:
    if (fd >= 0) {
	close(fd);
    }
    exit(1);
}

static const int geometry[][5] = {
    { 1440, 80, 2, 18,  512 },
    { 1232, 77, 2,  8, 1024 },
    { 1200, 80, 2, 15,  512 },
    {  720, 80, 2,  9,  512 },
    {  640, 80, 2,  8,  512 },
    { 0, 0, 0, 0 }
};

static int size_to_geometry(int size, int *track, int *head, int *sector, int *block)
{
    int i;

    for (i = 0; geometry[i][0] != 0; i++) {
	if (geometry[i][0] == size) {
	    *track = geometry[i][1];
	    *head = geometry[i][2];
	    *sector = geometry[i][3];
	    *block = geometry[i][4];
	    return 0;
	}
    }
    return 1;
}

static void format(const char *path, int size, int verify, int verbose, int force)
{
    char *sg = NULL;
    int fd = -1;
    int wp;
    struct ufi_capacities *capacities;
    int type;
    int track, head, sector, block, t, h;
    int mount_flags;
    int is_fdd;
    int need_reset = force;
    char *buf = NULL;
    char c;
    int i, n;
    time_t tt;

    if (verbose > 1) {
	printf("format on device=%s, size=%d\n", path, size);
    }
    if (!force) {
	if ((is_fdd = is_usb_fdd(path)) < 0) {
	    goto error;
	}
	if (!is_fdd) {
	    fprintf(stderr, "%s: device is not usb fdd\n", path);
	    goto error2;
	}
	sg = malloc(PATH_MAX);
	if (convert_to_sd_path(path, sg, PATH_MAX) < 0) {
	    sg = errno < 0 ? "/dev/sd* family" : NULL;
	    errno = errno < 0 ? -errno : errno;
	    goto error;
	}
	if (ext2fs_check_if_mounted(sg, &mount_flags) != 0) {
	    goto error;
	}
	if (mount_flags & EXT2_MF_MOUNTED) {
	    fprintf(stderr, "%s: device is mounted\n", sg);
	    goto error2;
	}
	if (convert_to_sg_path(path, sg, PATH_MAX) < 0) {
	    if (verbose > 1 && errno == -ENOENT) {
		fprintf(stderr, "\"modprobe sg\" might be needed to prepare SCSI pass-thru device.\n");
	    }
	    sg = errno < 0 ? "/dev/sg* family" : NULL;
	    errno = errno < 0 ? -errno : errno;
	    goto error;
	}
    } else {
	sg = strdup(path);
    }
    if ((fd = open(sg, O_RDWR | O_NONBLOCK)) < 0) {
	goto error;
    }
    if ((type = ufi_test_unit_ready(fd)) == UFI_ERROR && !force) {
	goto error;
    }
    if (!force) {
	if (type == UFI_NO_MEDIA) {
	    fprintf(stderr, "%s: no media\n", path);
	    goto error2;
	}
	if (ufi_mode_sense(fd, &wp) != UFI_GOOD) {
	    goto error;
	}
	if (wp == UFI_PROTECTED) {
	    fprintf(stderr, "%s: write protected\n", path);
	    goto error2;
	}
    }
    if (size == 0) {
	if ((capacities = ufi_read_format_capacities(fd, &type)) == NULL) {
	    goto error;
	}
	size = (capacities->blocks * capacities->block_size) / 1024;
	if (type != UFI_FORMATTED_MEDIA) {
	    need_reset = 1;
	}
	if (size_to_geometry(size, &track, &head, &sector, &block) != 0 || block != capacities->block_size) {
	    fprintf(stderr, "unable to figure out geometry for current media\n");
	    goto error2;
	}
    } else {
	if (size_to_geometry(size, &track, &head, &sector, &block) != 0) {
	    fprintf(stderr, "unable to figure out geometry for size=%d\n", size);
	    goto error2;
	}
	if (!force) {
	    if ((capacities = ufi_read_format_capacities(fd, &type)) == NULL) {
		goto error;
	    }
	    if (type == UFI_NO_MEDIA) {
		fprintf(stderr, "%s: no media\n", path);
		goto error2;
	    }
	    if (IS_2HD(capacities->blocks * capacities->block_size / 1024) != IS_2HD(size)) {
		fprintf(stderr, "%s: media type mismatch\n", path);
		goto error2;
	    }
	    need_reset = (type != UFI_FORMATTED_MEDIA) || (capacities->blocks * capacities->block_size / 1024 != size);
	}
    }
    if (verbose) {
	printf("geometry: track=%d, head=%d, sector=%d, block=%d\n", track, head, sector, block);
    }
    for (t = 0; t < track; t++) {
	for (h = 0; h < head; h++) {
	    if (verbose) {
	        printf("formatting track=%02d, head=%d\r", t, h);
		fflush(stdout);
	    }
	    if (ufi_format_unit(fd, track * head * sector, block, t, h) != UFI_GOOD) {
		goto error;
	    }
	}
    }
    if (need_reset) {
	if (ufi_reset(fd) != UFI_GOOD) {
	    goto error;
	}
    }
    close(fd);
    fd = -1;

    if (verify) {
	sg = realloc(sg, PATH_MAX);
	if (convert_to_sd_path(path, sg, PATH_MAX) < 0) {
	    sg = errno < 0 ? "/dev/sd* family" : NULL;
	    errno = errno < 0 ? -errno : errno;
	    goto error;
	}
	tt = time(NULL);
	// some devices need retry
	while ((fd = open(sg, O_RDONLY)) < 0 && errno == ENOMEDIUM && time(NULL) - tt < 10) {
	    sleep(1);
	}
	if (fd < 0) {
	    goto error;
	}
	buf = malloc(sector * block);
	for (t = 0; t < track; t++) {
	    for (h = 0; h < head; h++) {
		if (verbose) {
		    printf("verifying  track=%02d, head=%d\r", t, h);
		    fflush(stdout);
		}
		if ((n = read(fd, buf, sector * block)) != sector * block) {
		    if (n < 0) {
			goto error;
		    }
		    fprintf(stderr, "%s: short read from floppy disk on verification\n", sg);
		    goto error2;
		}
		if (t == 0 && h == 0) {
		    c = buf[0];  // normally 0xe5
		}
		for (i = 0; i < sector * block; i++) {
		    if (buf[i] != c) {
			fprintf(stderr, "%s: bad value in floppy disk on verification\n", sg);
			goto error2;
		    }
		}
	    }
	}
    }
    if (verbose) {
	printf("done                                   \n");
    }
    return;

 error:
    perror(sg == NULL ? path : sg);
 error2:
    if (fd >= 0) {
	close(fd);
    }
    exit(1);
}

static void usage(int exit_code)
{
    fputs(
	 "Usage: " PACKAGE " [OPTION]... [DEVICE]\n"
	 "Format a floppy disk in a USB floppy disk DEVICE.\n"
	 "\n"
	 "  -f, --format [SIZE]  specify format capacity SIZE in KB\n"
	 "                       without -f option, the format of the current media will be used\n"
	 "  -V, --verify         verify the medium after formatting\n"
	 "  -F, --force          do not perform any safety checks\n"
	 "  -i, --inquire        show device information, instead of performing format\n"
	 "                       without DEVICE argument, list USB floppy disk devices\n"
	 "  -v, --verbose        show detailed output\n"
	 "  -q, --quiet          suppress minor output\n"
	 "  -h, --help           show this message\n"
	 "\n",
	 exit_code == 0 ? stdout : stderr);
    exit(exit_code);
}

static const struct option long_options[] = {
    { "inquire", no_argument,       0, 'i' },
    { "format",  required_argument, 0, 'f' },
    { "verify",  no_argument,       0, 'V' },
    { "force",   no_argument,       0, 'F' },
    { "verbose", no_argument,       0, 'v' },
    { "quiet",   no_argument,       0, 'q' },
    { "help",    no_argument,       0, 'h' },
    { 0, 0, 0, 0 }
};

static const char short_options[] = "if:VFvqh";

int main(int argc, char **argv)
{
    int c;
    int opt_inquire = 0;
    int opt_format = 0;
    int opt_verify = 0;
    int opt_force = 0;
    int opt_verbose = 1;
    int format_size = 0;
    char tmp[256];

    while ((c = getopt_long(argc, argv, short_options, long_options, NULL)) != -1) {
        switch (c) {
        case 'i':
	    opt_inquire = 1;
            break;
        case 'f':
	    opt_format = 1;
	    format_size = atoi(optarg);
	    sprintf(tmp, "%d", format_size);
	    if (strcmp(optarg, tmp) != 0 || format_size < 0) {
		printf("format parameter should be numeric\n");
		usage(1);
	    }
            break;
        case 'V':
	    opt_verify = 1;
            break;
        case 'F':
	    opt_force = 1;
            break;
        case 'v':
	    opt_verbose = 2;
            break;
        case 'q':
	    opt_verbose = 0;
            break;
        case 'h':
	    usage(0);
	    break;
        case '?':
        default:
	    usage(1);
        }
    }

    if (!opt_inquire && !opt_format) {
	opt_format = 1;
	format_size = 0;
    }
    if (opt_format && opt_inquire) {
	fprintf(stderr, "%s: both inquire and format options specified\n", argv[0]);
	usage(1);
    } else if ((opt_format || opt_inquire) && argc - optind > 1) {
	fprintf(stderr, "%s: two or more devices specified\n", argv[0]);
	usage(1);
    } else if (opt_format && argc == optind) {
	fprintf(stderr, "%s: no device specified\n", argv[0]);
	usage(1);
    }

    ufi_set_verbose(opt_verbose);
    if (opt_inquire) {
	if (argc == optind) {
	    list_devices(opt_verbose);
	} else {
	    inquire(argv[optind], opt_verbose, opt_force);
	}
    } else if (opt_format) {
	format(argv[optind], format_size, opt_verify, opt_verbose, opt_force);
    }
    exit(0);
}
